Refactor appender to use the render function instead of a component instance

renatho opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

## Description
<!-- Please describe what you have changed or added -->

This deprecates the `renderAppender` prop in the `InnerBlocks`, introducing a new `appender` prop for this purpose. The difference is that the `appender` now is a rendered component instead of the component function/class.
If the `appender` is `null`, nothing is rendered.
If the `appender` is `undefined` the default is rendered.
More details [here](https://github.com/WordPress/gutenberg/pull/29911#discussion_r666217613).

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

- Tested running the tests, and in a real project using the feature.

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

It's basically a refactor that should not impact in anything.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/native-mobile.md -->

Commit:
Refactor appender to use the render funcition instead of a component instance

talldan commented:
I don't think this would work if renderAppender used a class component. So it'd be a breaking change.

talldan commented:
I've updated the docs to match how it currently works - https://github.com/WordPress/gutenberg/pull/29925.

renatho commented:
> I don't think this would work if renderAppender used a class component. So it'd be a breaking change.

Ah, that's true. I'm closing it, and thank you for updating the documentation!

EDIT: since we support the `class` too, I don't think it would be necessary the refactor. But if someone think it would be nice, I'd be happy to re-open this and add the support to classes, keeping the render prop option as the render prop pattern.

Commit:
Refactor to have backward compatibility

renatho commented:
Reopening this PR since we don't have a way to pass custom props to the appender unless using a function. I made a refactor introducing a new boolean property `renderAppenderAsFunction`, so it keeps the backward compatibility, and we can use the render prop by setting it to `true`.

I thought of using a separate prop for the render prop, but probably it would get more confusing.

renatho commented:
> Reopening this PR since we don't have a way to pass custom props to the appender unless using a function. I made a refactor introducing a new boolean property `renderAppenderAsFunction`, so it keeps the backward compatibility, and we can use the render prop by setting it to `true`.
> 
> I thought of using a separate prop for the render prop, but probably it would get more confusing.

Hi @talldan! Have you seen these changes?

While that, we're using `useCallback` as workaround, as it's used in some cases in Gutenberg. But I don't think the `useCallback` for this would be the best solution because we're trusting that to not re-create the instance of the component.

ellatrix commented:
If there's no issue with the current approach, is it worth adding an additional prop (public API) for this?

renatho commented:
> If there's no issue with the current approach, is it worth adding an additional prop (public API) for this?

I think the problem with the current approach is that in these cases we're trusting in the `useCallback` to create the component function that should be a static thing. As you said, it's not an issue now, but it could be in the future, depending on React improvements. 🤔 So if you think we could keep it as it is, and only fix it if we have a problem in the future, that's okay to me.

But since it doesn't seem a good practice, my opinion would be that we should introduce a solution for these cases (needing to use local props or states to create the appender). 😊 Some solution like this new prop would also make more transparent that the `renderAppender` is not a render prop.

talldan commented:
I can see the shortcomings with `renderAppender` and personally I'm ok with moving on from it if there's no good alternative. While I don't really like render props as a pattern, the current implementation does seem to have been a source of bugs for some plugins and the naming causes confusion.

If there's a new prop, I think it'd be best to deprecate `renderAppender`.

`renderAppenderAsFunction` is a bit verbose. What are the alternatives names? Could it just be `appender`? Or `renderBlockAppender`?

renatho commented:
> I can see the shortcomings with `renderAppender` and personally I'm ok with moving on from it if there's no good alternative. While I don't really like render props as a pattern, the current implementation does seem to have been a source of bugs for some plugins and the naming causes confusion.
> 
> If there's a new prop, I think it'd be best to deprecate `renderAppender`.
> 
> `renderAppenderAsFunction` is a bit verbose. What are the alternatives names? Could it just be `appender`? Or `renderBlockAppender`?

Hey there!

Honestly, I suggested the render prop because the current prop naming (`renderAppender`).

But I think the render prop is not needed here for now because we don't send any prop for the component. It could be helpful for the future if we want add arguments coming from the `BlockListAppender` though.

So maybe for now we could just deprecate the property as suggested, and just send a rendered component as `appender`. Like:

```js
const appender = <MyAppender { ...myProps } />

return <InnerBlocks appender={ appender } />
```

WDYT?

Commit:
Refactor InnerBlocks to use a new prop appender

The new prop is the rendered appender.
This commit also deprecates the renderAppender prop.

renatho commented on packages/block-editor/src/components/block-list-appender/index.js:
```diff
@@ -30,19 +31,26 @@ function BlockListAppender( {
-	if ( isLocked || CustomAppender === false ) {
+	if ( CustomAppender ) {
+		deprecated( 'renderAppender prop of the BlockListAppender component', {
+			since: '5.7',
```
Is it correct to use the `5.7` here?

renatho replied:
Also, I added the deprecation only in the component that implemented the changes, instead of all the components that receive this prop to avoid multiple warning deprecations.

talldan replied:
I think it should be the plugin version rather than the WordPress version. Best to go for the next one (10.8) and then if it misses it can be updated.

talldan replied:
My previous comment about the deprecation message applies here to:
> I think most implementers will specify this prop on the InnerBlocks component or useInnerBlocksProps hook. They probably won't be aware of BlockListAppender.

> It might be easiest to omit the component name from the message, or change the wording to 'renderAppender prop of InnerBlocks component or useInnerBlocksProps hook'.

renatho replied:
https://github.com/WordPress/gutenberg/pull/29911/commits/28f45bd315d7aa36af36cc4f36e009c9add38b20

renatho reviewed (commented):
I implemented the new approach. I think deprecating the `renderAppender` and making the `appender` be the new prop, as suggested, made it much cleaner. =)

Commit:
Update to show warning also when setting false to the renderAppender

Commit:
Update appender documentation

renatho commented on packages/block-editor/src/components/inner-blocks/README.md:
```diff
@@ -131,36 +131,36 @@ If locking is not set in an `InnerBlocks` area: the locking of the parent `Inner
 If the block is a top level block: the locking of the Custom Post Type is used.
 
-### `renderAppender`
+### `appender`
```
I also took the opportunity to update the docs.

Commit:
Refactor tests and components using renderAppender

Commit:
Refactor React Native InnerBlocks to use appender prop

renatho commented:
I refactored the React Native components too because it also contained the `renderAppender`. https://github.com/WordPress/gutenberg/pull/29911/commits/f89495942d2530a525a5e380a9e0cfd34042179f

I don't have the env setup to test the React Native env, so I'd really appreciate if someone could test this part for me. It seems safe, but it's worth a test. =)

Commit:
Update to support appender false in react native

talldan commented on packages/block-library/src/navigation-link/edit.js:
```diff
@@ -410,14 +410,16 @@ export default function NavigationLinkEdit( {
-					? InnerBlocks.DefaultAppender
-					: false,
+				hasDescendants ? (
+					<InnerBlocks.DefaultAppender />
```
Does this work ok? I thought it was `InnerBlocks.DefaultBlockAppender`. It might have been resolving as `undefined` all this time. 😄

renatho replied:
Good catch! Fixed here: https://github.com/WordPress/gutenberg/pull/29911/commits/67c2525c25dceef43900aeaf8fedf8b2a4d758a0

talldan commented on packages/block-library/src/navigation/edit.js:
```diff
@@ -85,12 +85,14 @@ function Navigation( {
-					? InnerBlocks.DefaultAppender
-					: false,
+				isSelected ? (
+					<InnerBlocks.DefaultAppender />
```
Same here with the name of the component.

renatho replied:
Also fixed here: https://github.com/WordPress/gutenberg/commit/67c2525c25dceef43900aeaf8fedf8b2a4d758a0

talldan commented on packages/block-editor/src/components/block-list-appender/index.native.js:
```diff
@@ -22,12 +23,24 @@ function BlockListAppender( {
 } ) {
-	if ( isLocked ) {
+	if ( undefined !== CustomAppender ) {
+		deprecated( 'renderAppender prop of the BlockListAppender component', {
+			since: '5.7',
+			alternative: 'appender prop of the BlockListAppender component',
+		} );
```
I think most implementers will specify this prop on the `InnerBlocks` component or `useInnerBlocksProps` hook. They probably won't be aware of `BlockListAppender`.

It might be easiest to omit the component name from the message, or change the wording to 'renderAppender prop of InnerBlocks component or useInnerBlocksProps hook'.

talldan replied:
Or alternatively could list all the components that this effects, e.g.
''renderAppender prop of the BlockListAppender and InnerBlocks components, and the useInnerBlocksProps hook'

renatho replied:
Makes sense! I decided to just omit the component, otherwise, it would demand the message maintenance depending on the components hierarchy changes. ;)
https://github.com/WordPress/gutenberg/pull/29911/commits/28f45bd315d7aa36af36cc4f36e009c9add38b20

renatho replied:
Also added a changelog entry here: https://github.com/WordPress/gutenberg/pull/29911/commits/44f6fa5ee4e98f13a15a73e2bff52e20214ee395

talldan reviewed (commented):
This looks pretty tidy, good job so far!

Commit:
Fix defaul block appender component

Commit:
Update renderAppender deprecation message

renatho commented:
> This looks pretty tidy, good job so far!

@talldan, Thanks a lot for your very valuable review! I just applied the changes. =)
Anything else, let me know.

Commit:
Add renderAppender deprecation to changelog

renatho commented:
It seems there are still some e2e issues. I couldn't run it locally yet because some errors were happening to me and I didn't dig into it to check what happened. I'll try to do it some time to fix the errors that are still happening. ;)

Commit:
Fix InnerBlocks appender tests

Commit:
Update to reproduce the previous behavior

renatho commented on packages/block-library/src/navigation/edit.js:
```diff
@@ -85,11 +84,11 @@ function Navigation( {
 				( isImmediateParentOfSelectedBlock &&
 					! selectedBlockHasDescendants ) ||
 				isSelected
-					? InnerBlocks.DefaultAppender
```
If we update it to the correct `DefaultBlockAppender`, it doesn't get the expected behavior (only the plus button), so I updated it to `undefined`, keeping the same previous behavior.

renatho replied:
If we update it to the correct `DefaultBlockAppender`, it doesn't get the expected behavior (only the plus button), so I updated it to `undefined`, keeping the same previous behavior.

renatho reviewed (commented):
Folks, thank you for your reviews so far!
Sorry that I didn't do it previously, but the e2e tests are also fixed now. 🙂

renatho commented:
Is there something I could do to go ahead with this fix? 🙂

Commit:
Update deprecation version

renatho commented:
Hey folks! Since the last time, there were some changes that conflicted with this PR. I fixed these conflicts now and tested it again. if someone could review it before we have more conflicts, I'd appreciate that. 😊

cc @talldan @ellatrix

kevin940726 commented on packages/block-editor/src/components/block-list-appender/index.js:
```diff
@@ -30,19 +31,26 @@ function BlockListAppender( {
+		} );
+	}
+
+	if ( isLocked || appender === false || CustomAppender === false ) {
```
Could we take this chance to deprecate `false` as a value too? This always seem a little bit weird to me while we can always just use `null` for the same thing.

renatho replied:
Hey Kevin! Thank you for checking!

I agree with you, but I think it would be a bigger breaking change. It would break sites that currently use `null` because any falsy (except the `false`) value runs the `else`. So I think it could be done in a future big change (maybe when changing this component for something else?). WDYT?

renatho replied:
Thank you for checking! =)
Hum. This makes sense! I'll take a look in the next few hours!

renatho replied:
This was implemented here: https://github.com/WordPress/gutenberg/pull/29911/commits/2537416fac2a8aba94a6160cc78a7caf16b4b453

I also added to the PR description, but copying the details here:

If the `appender` is `null`, nothing is rendered.
If the `appender` is `undefined` the default is rendered.
If the `appender` has any other value, it will be rendered.

Thank you again for the suggestion @kevin940726 !

renatho replied:
Just adding that I thought to not render anything for any falsy value instead of `null` only, but it wouldn't make sense for `0`, for example.
For other falsy (except `undefined` and `null`) it will render the `<TagName>` with the falsy value inside (the React handles if it should be rendered or not).

Commit:
Refactor to use fallback only when appender is `null`

kevin940726 commented on packages/block-editor/src/components/inner-blocks/README.md:
```diff
@@ -131,36 +131,36 @@ If locking is not set in an `InnerBlocks` area: the locking of the parent `Inner
 
--   **Type:** `Component|false`
--   **Default:** - `undefined`. When `renderAppender` is not specified, the default appender is shown. If a `false` value is provided, no appender is rendered.
+-   **Type:** `Object|false`
```
I think it should be `Object|null` now?

renatho replied:
Ah, yeah! Good catch! Amended here: https://github.com/WordPress/gutenberg/pull/29911/commits/463ab85a5ca460d5e8fbd58dbe4bbd9b0a59ea8e

kevin940726 commented on packages/block-library/src/column/edit.native.js:
```diff
@@ -111,30 +111,22 @@ function ColumnEdit( {
-	}, [ contentStyle[ clientId ], screenWidth, isSelected, hasChildren ] );
+	const { width: columnWidth } = contentStyle[ clientId ];
+	const isFullWidth = columnWidth === screenWidth;
+	const appender = isSelected && (
```
I think this is not the same as the original? We should be explicitly return `null` if `isSelected` is falsy?

renatho replied:
It would work because the implementation is a bit different in the RN component (rendering the falsy values except the `undefined`), but yeah, makes sense to use `null` to follow the documentation and the same contract as the non-RN implementation. =)
Good catch! Also fixed and amended in the same commit: https://github.com/WordPress/gutenberg/pull/29911/commits/463ab85a5ca460d5e8fbd58dbe4bbd9b0a59ea8e

renatho replied:
Thinking more, I refactored the behavior on the non-RN component to keep the same behavior for both components.

So as the opposite to what [commented previously](https://github.com/WordPress/gutenberg/pull/29911#discussion_r664890605), not when the appender contains any value different of `undefined` that is falsy, it will return null for both cases.

https://github.com/WordPress/gutenberg/pull/29911/commits/70235deb80b80ac0b8cfce57c5abf16b4a2e5d2e

WDYT? In the documentation, I think we can keep only the `null` as recommended value to not render anything.

kevin940726 replied:
Sure :). Sounds good to me.

renatho replied:
Just to let you know that it was changed again here: https://github.com/WordPress/gutenberg/pull/29911#discussion_r669365482 It seems to make more sense now =)

kevin940726 commented on packages/block-library/src/columns/edit.native.js:
```diff
@@ -131,18 +131,11 @@ function ColumnsEditContainer( {
-		}
-		return null;
-	};
+	const appender = isSelected && (
```
Same here.

renatho replied:
https://github.com/WordPress/gutenberg/pull/29911#discussion_r666208140

kevin940726 commented on packages/block-library/src/group/edit.native.js:
```diff
@@ -91,7 +87,7 @@ function GroupEdit( {
 			{ resizeObserver }
 			<InnerBlocks
-				renderAppender={ isSelected && renderAppender }
+				appender={ isSelected && appender }
```
Same here.

renatho replied:
https://github.com/WordPress/gutenberg/pull/29911#discussion_r666208140

kevin940726 reviewed (commented):
Mostly look good to me!

Commit:
Refactor appender to return null with any falsy value but undefined

Commit:
Fix new component appender prop

renatho commented:
Rebased and fixed a new case: https://github.com/WordPress/gutenberg/pull/29911/commits/5ec39b56e92d5bccd7bbe37fa0818d514c6af2f7 😉

talldan commented on packages/block-editor/src/components/block-list-appender/index.js:
```diff
@@ -30,19 +31,32 @@ function BlockListAppender( {
 	tagName: TagName = 'div',
 } ) {
-	if ( isLocked || CustomAppender === false ) {
+	if ( undefined !== CustomAppender ) {
```
Not that it makes a big difference, but the preference in Gutenberg's JavaScript is not to use yoda conditions, because the linting here catches any issues.

renatho replied:
Ah, perfect! Updated here: https://github.com/WordPress/gutenberg/pull/29911/commits/b2d0defbe249855354ad29b598bbd4d99553ef88.
TIL 😀

talldan commented on packages/block-editor/src/components/block-list-appender/index.js:
```diff
@@ -30,19 +31,32 @@ function BlockListAppender( {
+
+	if (
+		isLocked ||
+		( undefined !== appender && ! appender ) ||
```
It's not undefined, but it is falsey, so I imagine this is to check if it's `null` or `false`? e.g. values that people often use to render nothing in React.

Maybe it could be `( appender === null || appender === false )` to be more explicit.

edit: I've now seen the earlier comment about removing support for `false`, so it could maybe just be `isLocked || appender === null || CustomAppender === false`. I'm still ok with handling `false` as a corner case though.

renatho replied:
Sure! It makes more sense! https://github.com/WordPress/gutenberg/pull/29911/commits/25317c9352acca0c96e7ef004ef721616a5f8bda

talldan commented on packages/block-editor/src/components/block-list-appender/index.js:
```diff
@@ -30,19 +31,32 @@ function BlockListAppender( {
-		// Prefer custom render prop if provided.
-		appender = <CustomAppender />;
+	let blockAppender;
+	if ( undefined !== appender ) {
```
As other falsey values are avoided by the earlier `if` statement, this could probably just be `if ( appender ) {`.

What do you think?

renatho replied:
Ah, yeah! Makes sense too! https://github.com/WordPress/gutenberg/pull/29911/commits/25317c9352acca0c96e7ef004ef721616a5f8bda

talldan commented on packages/block-editor/src/components/inner-blocks/README.md:
```diff
@@ -131,36 +131,36 @@ If locking is not set in an `InnerBlocks` area: the locking of the parent `Inner
+-   **Default:** - `undefined`. When `appender` is not specified, the default appender is shown. If a `null` value is provided, no appender is rendered.
 
-A component to show as the trailing appender for the inner blocks list.
+A instantiated component to show as the trailing appender for the inner blocks list.
```
Small typo, it should be 'An instantiated component'. Could also use the term 'A react element' here.

renatho replied:
Oops! Also fixed in this commit: https://github.com/WordPress/gutenberg/pull/29911/commits/25317c9352acca0c96e7ef004ef721616a5f8bda

talldan commented on packages/block-editor/src/components/inner-blocks/README.md:
```diff
@@ -131,36 +131,36 @@ If locking is not set in an `InnerBlocks` area: the locking of the parent `Inner
-    -   `InnerBlocks.DefaultBlockAppender` - display the default block appender, typically the paragraph style appender when the paragraph block is allowed.
+    -   `<InnerBlocks.ButtonBlockAppender />` - display a `+` (plus) icon button as the appender.
+    -   `<InnerBlocks.DefaultBlockAppender />` - display the default block appender, typically the paragraph style appender when the paragraph block is allowed.
 -   Consumers are also free to pass any valid component. This provides the full flexibility to define a bespoke block appender.
```
I think this line will need to be updated to say 'any valid react element' or 'any instantiated component'.

renatho replied:
Also in https://github.com/WordPress/gutenberg/pull/29911/commits/25317c9352acca0c96e7ef004ef721616a5f8bda

talldan reviewed (commented):
@renatho Thanks for your patience, and apologies I disappeared for a bit, ended up being very focused on WordPress 5.8 issues.

I could only spot a few minor things. I'll also ask if one of the regular contributors to the native codebase can give this a test.

Commit:
Revert yoda contitions

Commit:
Update appender checks

Commit:
Simplify some checks

Appender explicitly supports `false` as an option now.

Commit:
Update deprecated version

renatho commented:
> @renatho Thanks for your patience, and apologies I disappeared for a bit, ended up being very focused on WordPress 5.8 issues.

Ah, no problem! Thank you for the review, @talldan!!
I added some commits and answered your comments. I also updated the deprecated version here: https://github.com/WordPress/gutenberg/pull/29911/commits/1ce82df41188fd02cb88ed01b544a993a0ed8258

ceyhun commented on packages/block-library/src/column/edit.native.js:
```diff
@@ -111,30 +111,22 @@ function ColumnEdit( {
-		}
-		return null;
-	}, [ contentStyle[ clientId ], screenWidth, isSelected, hasChildren ] );
+	const { width: columnWidth } = contentStyle[ clientId ];
```
Could this be removing the performance optimization previously made here by using `useCallback`? I remember there were a few performance optimization done around here before: https://github.com/WordPress/gutenberg/pulls?q=is%3Apr+sort%3Aupdated-desc+performance+label%3A%22Mobile+App+-+i.e.+Android+or+iOS%22+is%3Amerged+label%3A%22%5BType%5D+Performance%22

Could maybe using `useRef` bring that back?

renatho replied:
Hey @ceyhun!

Thank you for checking and testing! =)

I imagine the reason for the performance issues was probably that it was creating a new function component for every render (the main problem of the previous approach of `renderAppender`). It was avoiding any optimization from React reconciliation. Now it should be fine.
I'd suggest keeping it as it is, and if we really see new performance issues, we could try to fix the origin of this. WDYT?

ceyhun replied:
I think now it will return a new `appender` component for every render instead. But I guess that was also happening before, even when the same `renderAppender` function was called. In that case I agree to keep it as it is.

ceyhun commented on packages/block-library/src/group/edit.native.js:
```diff
@@ -39,23 +38,20 @@ function GroupEdit( {
-			</View>
-		),
-		[ align, hasInnerBlocks ]
+	const appender = (
```
Same as https://github.com/WordPress/gutenberg/pull/29911/files?file-filters%5B%5D=.js&file-filters%5B%5D=.php&file-filters%5B%5D=.snap#r669839983

ceyhun reviewed (commented):
I've tested this on iOS and Android using Column and Group blocks and it's looking good. I just have a comment about possible loss of a previous performance improvement.

renatho commented:
@talldan, I think after the mobile review, we're good to go ahead with this. WDYT?

Commit:
Update deprecation versions

Commit:
Update components to use appender prop instead of deprecated renderAppender

renatho commented:
Rebased to fix the new conflicts, and updated the deprecated versions! ;)

Mamaduka commented:
I'm doing some old PR cleanup. Since there has been no activity in the last four years, I'm going to close this as "stale."
